components/tidb_query_datatype/src/codec/mysql/json/path_expr.rs

```diff
@@ -73,82 +65,331 @@ impl PathExpression {
+                }
+
+                let key = unsafe { self.input.get_unchecked(key_start..key_end) }.to_string();
+
+                // it's not quoted, we'll have to validate whether it's an available ECMEScript
+                // identifier
+                for (i, c) in key.char_indices() {
+                    if i == 0 && c.is_ascii_digit() {
+                        return Err(box_json_path_err!(key_start + i));
+                    }
+                    if !c.is_ascii_alphanumeric() && c != '_' && c != '$' && c.is_ascii() {
+                        return Err(box_json_path_err!(key_start + i));
+                    }
+                }
+
+                Ok(PathExpressionToken::Leg((
+                    PathLeg::Key(key),
+                    Position {
+                        start,
+                        end: key_end,
+                    },
+                )))
+            }
+        }
+    }
+
+    fn next_index(&mut self) -> Result<PathExpressionToken> {
+        let (start, _) = self.char_iterator.next().unwrap();
+
+        self.trim_white_spaces();
+        if self.reached_end() {
+            return Err(box_json_path_err!(self.current_index()));
         }
+
+        return match self.char_iterator.next().unwrap() {
+            (_, '*') => {
+                // then it's a glob array index
+                self.trim_white_spaces();
+                if self.reached_end() {
+                    return Err(box_json_path_err!(self.current_index()));
+                }
+
+                if self.char_iterator.next_if(|(_, ch)| *ch == ']').is_none() {
+                    return Err(box_json_path_err!(self.current_index()));
+                }
+
+                Ok(PathExpressionToken::Leg((
+                    PathLeg::Index(PATH_EXPR_ARRAY_INDEX_ASTERISK),
+                    Position {
+                        start,
+                        end: self.current_index(),
+                    },
+                )))
+            }
+            (number_start, '0'..='9') => {
+                // then it's a number array index
+                while self
+                    .char_iterator
+                    .next_if(|(_, ch)| ch.is_ascii_digit())
+                    .is_some()
+                {}
+                let number_end = self.current_index();
+
+                self.trim_white_spaces();
+                // now, it reaches the end of input, or reaches a non-digit character
+                match self.char_iterator.peek() {
+                    Some((_, ']')) => {}
+                    Some((pos, _)) => {
+                        return Err(box_json_path_err!(pos));
+                    }
+                    None => {
+                        return Err(box_json_path_err!(self.current_index()));
+                    }
+                }
+                self.char_iterator.next().unwrap();
+
+                let index = self.input[number_start..number_end].parse::<i32>().unwrap();
```

Comment:
If it overflows, it will make TiKV panic?

Reply:
And I'm not sure what the max acceptable index should be. I try MySQL 5.7.39 and 8.0.30 in docker, the max allowed index is `2^63-1` and `2^32-1` respectively...

Reply:
Nice catch! But I didn't modify the data type, so the upper limit of current implementation is still `2^31-1`. We are using signed type here because we use the `-1` as a special digit represent `*`. It should be refractored to be an `enum {Asterisk, Index(u64)}`. I'll add an issue for it.